scania-digital-design-system/tegel
# feat(tegel-angular): reactive forms integration

**Describe pull-request**  
Adds native integration with the angular reactive forms for the Tegel input components.

**Solving issue**  
Fixes: #546

**How to test**  
This one is a bit tricky to test locally since it requires packages to be packed locally first.

1. Navigate to `packages/core`
2. Create a `packages` directory in your home folder
3. Run `npm run build`
4. Run `npm pack --pack-destination ~/packages` - This should create `$homeFolder/packages/scania-tegel-1.8.0.tgz`
5. Update the @scania/tegel package in packages/angular/package.json to "@scania/tegel": "file:~/packages/scania-tegel-1.8.0.tgz",
6. run `npm run build` inside packages/angular
7. Run `npm pack --pack-destination ~/packages` inside packages/angular
8. Open your angular project (I used the Tegel demo app) and update the `@scania/tegel-angular` package to `"@scania/tegel-angular": "file:~/packages/scania-tegel-angular-1.8.0.tgz"`
9. Remove all `ngDefaultControl` directives from your inputs.
10. Run the dev server
11. Test out the forms.

**Checklist before submission**

**Suggested test steps**

**Screenshots**  

**Additional context**  
The existing `angularValueAccessorBindings` configuration inside stencil.config.ts was mostly correct, but it was not applied, since the generated value accessors weren't exported from the `TegelModule`. The value accessors are now exported. `tds-dropdown` did not work with the auto-generated value accessors out of the box, so I needed to implement an "adapter" value accessor (packages/angular/projects/components/src/lib/directives/tds-dropdown-value-accessor.ts)


The `handleInput` event in `tds-text-field` and `tds-textarea` was emitting the event before updating its value. This caused issues when binding to tdsInput, since `event.target.value` was the previous value and not the current value. I fixed this by changing the order inside the method from:
```
handleInput(event): void {
    this.tdsInput.emit(event);
    this.value = event.target.value;
  }
```
To:
```
handleInput(event): void {
    this.value = event.target.value;
    this.tdsInput.emit(event);
  }
```

Forms integration has been implemented for the following components:
- [x] tds-checkbox
- [x] tds-chip[type="checkbox"]
- [x] tds-chip[type="radio"]
- [x] tds-radio-button
- [x] tds-dropdown
- [x] tds-text-field
- [x] tds-textarea
- [x] tds-datetime
- [x] tds-slider
- [x] tds-toggle
